Fix statement vs expr

jakzale opened the pull request:
Fixes #51, #54 

Essentially we make `[Statement f]` as our program and the main parsing rule.

- Fix `Sign` vs `Sig` naming issue,
- add `Statements f` newtype for `[Statement f]`
- add `Program f` newtype of `Statements f` as top-level node of AST.
- duplicate necessary functions with
  - `stmt` prefix for statements, and
  - `expr` prefix for expressions.
- update and clean up all the test cases.
- add `stack bench` for testing the generators.

In addition we add the benchmark of quickcheck generators

effectfully commented on test/Field/golden/forLoops.golden:
```diff
@@ -1 +1,16 @@
\ No newline at end of file
+let i'_6 = 1;
+let k_7 = 1 * 2;
+assert k_3 == (i'_1 * 2);
```
Much better, thanks.

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -211,15 +212,11 @@ denoteExpr env = fmap denoteUniConst . evalExprUni env
-    EStatement (EAssert exprN) <$> normExpr env rest
+    => Env (SomeUniConst f) -> Statement f -> m (Statement f)
+normStatement env (ELet uniVar def) = do
+    ELet uniVar <$> normExpr env def
```
That changes the behavior. Previously we would inline things like `let x = 3`, but now we don't. This is needed for shrinking, so we have to fix that.

effectfully commented on field/TinyLang/Field/Raw/Core.hs:
```diff
@@ -24,16 +26,24 @@ newtype Var = Var { unVar :: Identifier }
+data Statement v f
+    = ELet    v          (Expr v f)
+    | EAssert (Expr v f)
+    | EFor    v          Integer    Integer (Statements v f)
```
Why do we need to wrap `[Statement v f]` with a `newtype` here?

jakzale replied:
TL;DR: so we can reuse the operations that we define over `Statements`.  `[Statements f]` does not exist on its own, therefore I am not keen on it occuring in the AST.

I believe that `Statements f` is semanticaly different than `[Statement f]`. At the moment I expose the accessor, but I do not think that I should be doing it.  With our current AST `[Statement f]` should not be exposed as I think that treating it as a list of statements can lead to potential bugs.

There are some operations that will work on `Statement` and `Program`/`Statements`, but they will not necessarily work on when mapping them over  `[Statement f]`, these operations include:
- renaming, and
- normalisation.

For renaming, we need to get a fresh unique, and that operation cannot be perfomed independently over statements.

For normalisation, we need to have an option of inlining `let` bindings, which you also noticed in a different comment.

effectfully replied:
Ok, please inline these comments somewhere in the code.

I'm not against having a `newtype` wrapper there. I'm against having the same `newtype` wrapper both for the entire program and for the body of a loop. Clearly those are very different things (e.g. to rename a `Program` we need to call `supplyAtLeastFree` over the entire program, but when we rename the body of a loop, there's no such requirement).

So if we can have both `Program = ...` and `Statements = ...`, that will be better, I believe.

effectfully commented on field/TinyLang/Field/Raw/Core.hs:
```diff
@@ -24,16 +26,24 @@ newtype Var = Var { unVar :: Identifier }
+{-| @Statements v f$, @Statement v f@, and @Expr v f@ are parameterised
+  by the type of variable @v@.
+-}
+newtype Statements v f = Statements { unStatements :: [Statement v f] }
```
I'd call this `Program` and not use the wrapper in `For`. So that we can have `parseProgram`, `checkProgram`, etc.

effectfully commented on field/TinyLang/Field/Rename.hs:
```diff
@@ -28,11 +34,11 @@ renameVarM var = do
     defRen <- renameExprM def
-    withFreshenedVar var $ \varFr -> cont $ ELet (UniVar uni varFr) defRen
-withRenamedStatementM (EAssert expr)              cont = renameExprM expr >>= cont . EAssert
+    withFreshenedVar var $ \varFr -> pure $ ELet (UniVar uni varFr) defRen 
```
Could you also add a property test where we generate a program, rename it and evaluate and then check if the result matches with just evaluating the program, without the renaming. Such a test has to fail here. `withFreshenedVar` updates the local context of the continuation, and here you just immediately leave that context with `pure`, so all the statements get renamed in the old context, not in the new one.

jakzale replied:
Done, still need to add tests.

jakzale replied:
Is there an easy way to generate an environment to close the program?

effectfully replied:
See [these lines](https://github.com/effectfully/tiny-lang/blob/master/field/TinyLang/Field/Generator.hs#L444-L457).

effectfully commented on field/TinyLang/Field/Rename.hs:
```diff
@@ -8,9 +9,14 @@ import           TinyLang.Field.Typed.Core
 
 renameExpr :: MonadSupply m => Expr f a -> m (Expr f a)
 renameExpr expr = do
```
Let's remove that for now, we're mostly interested in renaming programs rather than standalone expressions.

effectfully replied:
Still think it's worth removing this function.

jakzale replied:
Done

effectfully commented on field/TinyLang/Field/Typed/Core.hs:
```diff
@@ -180,18 +197,21 @@ instance Eq f => Eq (UniVar f a) where
+    EAssert as1 == EAssert as2 =
+        as1 == as2
+    EFor (UniVar u1 v1) i1 j1 stmts1 == EFor (UniVar u2 v2) i2 j2 stmts2 =
+        withGeqUni u1 u2 False $ v1 == v2 && i1 == i2 && j1 == j2 && stmts1 == stmts2
```
I'd expect that we can drop `withGeqUni u1 u2 False $`, 'cause we know those two are the same. It might be a little dangerous, because we might eventually allow to bind variables of non-field types in `for` loops, but then we can make it `u1 == u2 && ...` (if we don't have an `Eq` instance for `Uni`, we should add it).

effectfully commented on field/TinyLang/Field/Typed/Parser.hs:
```diff
@@ -30,20 +32,23 @@ import qualified Data.Map.Strict                  as Map
-instance TextField f => IsString (Some (Expr f)) where
-    fromString = _scopedValue <$> fromString
+-- instance TextField f => IsString (Some (Expr f)) where
+--     fromString = _scopedValue <$> fromString
```
Something wrong with it?

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -209,17 +224,24 @@ denoteExpr
+    => Env (SomeUniConst f) -> Statement f -> m (Statement f)
+normStatement env (ELet uniVar def) = ELet uniVar <$> normExpr env def
+normStatement env (EAssert expr) = EAssert <$> normExpr env expr
+normStatement env (EFor uniVar start end stmts) = EFor uniVar start end <$> normStatements env stmts
```
Ideally, we should drop the loop if its body normalizes to an empty list of statements.

effectfully commented on field/TinyLang/Field/Rename.hs:
```diff
@@ -29,10 +37,18 @@ renameVarM var = do
 withRenamedStatementM :: Statement f -> (Statement f -> RenameM c) -> RenameM c
-withRenamedStatementM (ELet (UniVar uni var) def) cont = do
+withRenamedStatementM (ELet (UniVar uni var) def) kont = do
     defRen <- renameExprM def
```
Please add a comment here explaining that we rename `def` outside of `withFreshenedVar`, because `var` is not in scope in `def` (kind of obvious, but worth having a comment still).

effectfully commented on field/TinyLang/Field/Rename.hs:
```diff
@@ -29,10 +37,18 @@ renameVarM var = do
+withRenamedStatementM (EFor (UniVar uni var) start end stmts) kont = do
+    withFreshenedVar var $ \varFr ->
+        withRenamedStatementsM stmts $ \stmtsRen ->
+            kont $ EFor (UniVar uni varFr) start end stmtsRen
```
We need a comment here explaining that we do not have lexical scoping and that is why `kont` is called inside `withRenamedStatementsM stmts`.

effectfully commented on field/TinyLang/Field/Rename.hs:
```diff
@@ -1,25 +1,33 @@
-    ( renameExpr
+    ( renameProgram
+    , renameExpr
     ) where
```
We need a test that shows how `Unique`s get resolved in

```
let a = a;
let a = a;
for a = 1 to 2 do
    let a = a;
    let a = a;
end;
let a = a;
let a = a;
```

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -144,17 +148,29 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
+    pure $ env
+evalStatementUni env (EFor (UniVar _ var) start end stmts) = foldM iteration env [start .. end] where
+    iteration env' i = evalStatementsUni env'' stmts where
+        env'' = insertVar var (Some $ fromInteger i) env'
```
I'd probably just make it

```haskell
evalStatementUni
    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
    => Statement f ->  StateT (Env (SomeUniConst f)) m ()
```

and then you don't need to trace the state manually like you do here.

effectfully commented on field/TinyLang/Field/Typed/TypeChecker.hs:
```diff
@@ -216,30 +220,25 @@ checkExpr m = do
+    :: forall m f. (MonadTypeChecker m, TextField f)
+    => R.Program R.Var f -> m (T.Program f)
+checkProgram =
+    liftA T.Program . liftA T.Statements . mapM checkStatement . R.unStatements .  R.unProgram
```
`fmap (T.Program . T.Statements)` instead of `liftA T.Program . liftA T.Statements`?

effectfully replied:
What if we define

```
newtype Statements stmt = Statements [stmt]

newtype Program stmt = Program (Statements stmt)
```

? Then we can reuse `Statements` and `Program` for both the typed and raw ASTs and this definition will be just `traverse checkStatement`.

The downside is that if the ASTs get out of sync, then this won't work anymore. But I'm not sure how likely that is.

jakzale replied:
Hmm, but in which module should I put those?  `Field/Core.hs`?

jakzale replied:
Hmm, should we also parameterise those newtype by the field?

Therefore yielding something akin to
```haskell
newtype Statements stmt f = Statements [stmt f]
```

effectfully replied:
> Hmm, but in which module should I put those? `Field/Core.hs`?

Whatever you think is appropriate. Might be `Field.Raw.Core` as well, since the typed language depends on the raw one anyway.

> Hmm, should we also parameterise those newtype by the field?

I think it shouldn't be required? You just instantiate `stmt` with a particular `Statement` (either raw or typed one) and it's what contains the `f`.

jakzale commented:
@effectfully, I added a benchmark (`stack bench`) for testing the generators.

Does the following output looks good to you?
```
Requested runs:  1000
Requested size:  1000

Minimum depth = 0
Maximum depth = 15
Mean depth    = 3.168

Minimum number of nodes = 0
Maximum number of nodes = 1890
Mean number of nodes    = 565.964
```

effectfully commented:
> Does the following output looks good to you?

Yes. We'll need some more refined tool to output other interesting stats (how many `add`s, `mul`s etc there are, how many variables the program has, how many of them are mentioned at least twice, how many comparisons are done on "negative integers", etc).

We'll also need to see if it makes the compiler tests slower, 'cause they are already quite slow.

But for now looks good.

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -204,26 +213,37 @@ denoteUniConst (UniConst Vector v) = unAField . fromInteger . packPositiveAsc $
+                go (insertVar var (Some uniConst) env)  rest
+            -- Drop for loop with an empty body
+            EFor _ _ _ stmts | [] <- unStatements stmts ->
+                go env rest
```
Hm. Together with

```haskell
evalStatementUni (EFor (UniVar _ var) start end stmts) =
    mapM_ iteration [start .. end] where
    iteration i = do
        modify' $ insertVar var (Some $ fromInteger i)
        evalStatementsUni stmts
```

that may result in a well-scoped program becoming ill-scoped. I'm not sure if we want the loop variable to be in scope after the loop, but let's do something consistent: either unscope the loop variable once the loop finishes or change the value of the variable to `end + 1` (currently it's `end`)  and do `go (insertVar var <...> env) rest` here (where `<...>` is `end + 1` written in a suitable manner).

I'd define a standalone function instead of the `mapM_ iteration [start .. end] ` thing (in particular in order to make the final value of the loop variable `end + 1` rather than `end`).

jakzale replied:
@effectfully after our discussion off github, should we agree on the variable being bound to `max start end` after the loop?

effectfully replied:
Yes. I'm not quite satisfied by this, but I'm not satisfied by every other option more or less equally, so let's go with `max start end`.

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -204,26 +213,37 @@ denoteUniConst (UniConst Vector v) = unAField . fromInteger . packPositiveAsc $
+                go env rest
+            -- Drop EAssert True
+            EAssert (EConst (UniConst _ True)) ->
+                go env rest
```
We could also explode on `assert False`.

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -252,13 +272,17 @@ normExpr env (EAppBinOp op e1 e2) = do
+
+instStatements :: MonadEvalError f m
+    => Env (SomeUniConst f) -> Statements f -> m (Statements f)
+instStatements env = fmap mkStatements . mapM (instStatement env) . unStatements
```
Would `instStatements = traverse . instStatement` type check?

jakzale replied:
`traverse (instStatement env)` works.

effectfully commented on field/TinyLang/Field/Raw/Core.hs:
```diff
@@ -24,16 +33,49 @@ newtype Var = Var { unVar :: Identifier }
+type Program v f    = C.Program    (Statement v f)
+type Statements v f = C.Statements (Statement v f)
+
+mkProgram :: Statements v f -> Program v f
+mkProgram = C.Program
+
+unProgram :: Program v f -> Statements v f
+unProgram = C.unProgram
```
Defined just for the sake of better type inference?

jakzale replied:
Defined so I do not need to import `TinyLang.Field.Core as C` everywhere due to name clashes.

effectfully replied:
What if you just reexport `Program` (the constructor, not the type) and `unProgram` from `TinyLang.Field.Raw.Core`?

jakzale replied:
See https://gitlab.haskell.org/ghc/ghc/issues/8753

effectfully replied:
Have you tried using the `pattern` trick?

effectfully commented on field/TinyLang/Field/Rename.hs:
```diff
@@ -1,25 +1,34 @@
 
+renameProgram :: MonadSupply m => Program f -> m (Program f)
+renameProgram prog = do
+    let stmts = unProgram prog
```
I'd do `renameProgram (Program stmts) = <...>`, but no big deal.

effectfully commented on field/TinyLang/Field/Rename.hs:
```diff
@@ -29,10 +38,23 @@ renameVarM var = do
+withRenamedStatementsM :: Statements f -> (Statements f -> RenameM c) -> RenameM c
+withRenamedStatementsM stmts kont =
+    runContT (traverse (ContT . withRenamedStatementM) stmts') $ kont . mkStatements where
+    stmts' = unStatements stmts
```
Same here, pattern matching on `stmts` directly feels more readable.

effectfully commented on field/TinyLang/Field/Typed/Core.hs:
```diff
@@ -228,32 +257,55 @@ isTracked uniq x env =
-supplyFromAtLeastFree :: MonadSupply m => Expr f a -> m ()
-supplyFromAtLeastFree =
-    supplyFromAtLeast . freeUniqueIntMap . unEnv . _scopedVarSignsFree . exprVarSigns
+stmtVarSigs' :: ScopedVarSigs f -> Statement f -> ScopedVarSigs f
```
I liked this function more when it was local. I don't think we need `exprSupplyFromAtLeastFree` and thus don't need `exprVarSigs` and thus can keep `stmtVarSigs'` and `exprVarSigs'` local just not to pollute the top-level with such names.

It might turn out that we do need `exprVarSigs`, in which case I'll return what you've written here.

jakzale replied:
Cleaned up the code now.

effectfully commented on field/TinyLang/Field/Typed/Parser.hs:
```diff
@@ -26,24 +30,24 @@ import qualified Data.IntMap.Strict               as IntMap
 parseScopedExpr
     :: forall f m. (MonadError String m, MonadSupply m, TextField f)
     => String -> m (Scoped (SomeUniExpr f))
 parseScopedExpr str = do
```
Let's drop this one as well and move the comment to `parseScopedProgram`.

jakzale replied:
Done

effectfully commented on field/TinyLang/Field/Typed/Parser.hs:
```diff
@@ -53,3 +57,31 @@ parseExpr
 parseExpr = fmap _scopedValue . parseScopedExpr
+
+
+parseScopedProgram'
```
Let's call it `parseScopedProgramFrom` or something, I don't like `'` in top-level names (unless they're about strictness and apart from a few exceptions like `Lens'`).

effectfully commented on field/TinyLang/Field/Typed/TypeChecker.hs:
```diff
@@ -216,30 +220,25 @@ checkExpr m = do
-    let env = insertVar var (Some $ fromInteger i) mempty
-        report err = error $ "Panic: " ++ show err
-    map (either report id . instStatement env) stats
+    T.EFor (T.UniVar Field tVar) start end . T.mkStatements <$> mapM checkStatement (R.unStatements stmts)
```
`T.EFor (T.UniVar Field tVar) start end <$> traverse checkStatement stmts`?

effectfully commented on test/Field/Textual.hs:
```diff
@@ -26,70 +27,73 @@ import           Test.Tasty.QuickCheck
 forgetStatementIDs (EAssert expr) = EAssert $ forgetIDs expr
+forgetStatementIDs (EFor uvar start end stmts) =
+    EFor (forgetID uvar) start end stmts' where
+    stmts' = forgetStatementIDs <$> stmts
```
We have this

```haskell
    expr where
    binding1 = ...
    ...
    bindingn = ...
```

in a number of places. I think it would be more readable to indent everything that follows after a `where` like that:

```haskell
    expr where
        binding1 = ...
        ...
        bindingn = ...
```

But not a big deal.

effectfully reviewed (commented):
I'll review the generators stuff a bit later.

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -144,22 +145,33 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
+
+evalStatementsUni :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
+    => Statements f -> StateT (Env (SomeUniConst f)) m ()
+evalStatementsUni = mapM_ evalStatementUni . unStatements
```
`evalStatementsUni = traverse_ evalStatementUni`?

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -228,81 +227,126 @@ arbitraryUnOpRing = elements [Neg]
+                     )
+
+-- TODO:  Is there a way to automate this?
+instance MonadState s m => MonadState s (GenT m) where
```
You could define a newtype, provide an instance for it and then use `DerivingVia`. But who cares.

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -228,81 +227,126 @@ arbitraryUnOpRing = elements [Neg]
+
+-- Not sure if this is the right type
+runSGenT :: (Monad m) => Vars f -> SGenT f m a -> Gen (m a)
+runSGenT vars = fmap ((flip evalStateT) vars) . fmap (runSupplyT) . runGenT . unSGenT
```
Normally, I'd make it

```haskell
runSGenT :: Monad m => Vars f -> SGenT f m a -> GenT m a
```

by using [`hoist`](http://hackage.haskell.org/package/mmorph-1.1.3/docs/Control-Monad-Morph.html#v:hoist), but `GenT` doesn't have an `MFunctor` instance, but you could provide one. Not a big deal.

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -228,81 +227,126 @@ arbitraryUnOpRing = elements [Neg]
+                               modify' (Some uniVar :)
+                               let size' = size - 1
+                                   start = arbitraryM
+                                   end   = arbitraryM
```
We shouldn't generate loops with arbitrary bounds, that's too much to compute. Let's keep loops within, say, 10 iterations.

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -228,81 +227,126 @@ arbitraryUnOpRing = elements [Neg]
+                               x <- boundedArbitraryExpr vars size'
+                               pure $ EAssert $ EAppBinOp binOp x x)
+                     -- generate a for loop
+                     , (3, do
```
I think loops should be generated rarely, perhaps 1 would be fine.

jakzale replied:
At the moment loops are the only construct that increases the depth of the AST.  I will play around with reducing the probability.

jakzale replied:
Reducing `for` loop generation to `1` still gives nice depth.

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -384,10 +421,34 @@ defaultUniConst =
+
+-- A modified shrinkList, that preserves the structure of the underlying list
+shrinkList' :: (a -> [a]) -> [a] -> [[a]]
+shrinkList' shr = shrinkOne where
```
Let's call it `shrinkElements`.

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -384,10 +421,34 @@ defaultUniConst =
+            adjustUniquesForVars vars
+            boundedArbitraryStmts size
+
+    shrink stmts = mkStatements <$> preservingShrink stmts' ++ nonPreservingShrink stmts' where
```
Oh, I didn't know the precedence of `<$>` was lower than the one of `++`. Please make it

```haskell
map mkStatements $ preservingShrink stmts' ++ nonPreservingShrink stmts'
```

just so I (and maybe someone else) won't get confused again.

jakzale replied:
replaces `++`s with concat.

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -384,10 +421,34 @@ defaultUniConst =
+
+    shrink stmts = mkStatements <$> preservingShrink stmts' ++ nonPreservingShrink stmts' where
+        stmts' :: [Statement f]
+        stmts' = unStatements stmts
```
Again would be nice to use simple pattern matching instead of introducing another binding.

jakzale replied:
Hmm, actually
`shrink stmts | stmts' <- unStatements stmts = .... where`
does not bind `stmts'` in the `where` clause, which I find surprising.

Wonder if that is on purpose...

effectfully replied:
You can use [`ViewPatterns`](https://gitlab.haskell.org/ghc/ghc/-/wikis/view-patterns) and write

```haskell
shrink (unStatements -> stmts) = ...
```

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -384,10 +421,34 @@ defaultUniConst =
     where
         uni = knownUni @f @a
 
--- We do not provide an implementation for 'arbitrary' (because we don't need it and it'd be
--- annoying to write it), but we still want to make provide an 'Arbitrary' instance, so that
--- 'shrink' can be used in the 'Arbitrary' instance of 'Expr' (a separately provided
--- 'shrinkStatement' wouldn't work, because we want to shrink a pair of values, see the instance).
```
The comment is still relevant for the `Arbitrary (Statement f)` instance.

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -396,7 +457,9 @@ instance (Field f, Arbitrary f) => Arbitrary (Statement f) where
     -- @lhs@ and @rhs@ respectively (just to have some shrinking that does not break the assertion).
-    shrink (EAssert expr)    = EAssert <$> shrink expr
+    shrink (EAssert expr) = EAssert <$> shrink expr
+    shrink (EFor uniVar start end stmts) =  efor <$> shrink (start, end, stmts) where
```
Shrinking `start` makes the loop more expensive to compute... Seems like if we shrink `start` to `start'`, then we should also subtract `start - start'` from `end`.

jakzale replied:
Hmm, I need to think how would that work in the presence of permitting normalisable expressions in the for loop.

jakzale replied:
Not necessarily, see my comment regarding weird loop ranges.

jakzale replied:
I do not think this is a problem now.  I made a note, we can revisit it we notice that there are massive performance degradations for running tests.

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -420,42 +483,39 @@ instance (KnownUni f a, Field f, Arbitrary f) => Arbitrary (Expr f a) where
-genEnvFromVarSigns =
-    traverse $ \(VarSign _ (uni :: Uni f a)) ->
+-- TODO:  Uncomment
+-- instance (Field f, Arbitrary f) => Arbitrary (SomeUniExpr f) where
```
Just delete it, I don't think we need it. Or do we?

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -420,42 +483,39 @@ instance (KnownUni f a, Field f, Arbitrary f) => Arbitrary (Expr f a) where
-        flip map (shrink someUniExpr) $ \shrunk@(SomeOf _ expr) ->
-            ExprWithEnv shrunk . Env . IntMap.intersection vals . unEnv $ exprFreeVarSigns expr
+-- TODO:  Uncomment
+-- instance (Field f, Arbitrary f) => Arbitrary (ExprWithEnv f) where
```
This should change to `ProgramWithEnv`, defined in a similar way.

jakzale replied:
Done

jakzale commented:
There are some issues with `ProgramWithEnv` arbitrary instances:
```
        *** Failed! (after 4 tests and 1 shrink):
        evaluation of renamed program
        [Some 16,Some (UniConst Bool False),Some 1]
        differs from evaluation of original program
        [Some (UniConst Bool False),Some 16]
        for final state of renamed program
        Right (Env {unEnv = fromList [(0,Some 16),(9,Some (UniConst Bool False)),(23,Some 1)]})
        for final state of evaluated program
        Right (Env {unEnv = fromList [(9,Some (UniConst Bool False)),(23,Some 16)]})
        for renamed program
        let u_0 = 16;
        
        for original program
        let u_23 = 16;
        
        ProgramWithEnv (Program (Statements [ELet (UniVar {_uniVarUni = Field, _uniVarVar = u_23}) (EConst 16)])) (Env {unEnv = fromList [(9,Some (UniConst Bool False)),(23,Some 1)]})

```

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -144,22 +146,37 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
+    unless exprR $ throwError $ AssertionFailedEvalError expr
+evalStatementUni (EFor (UniVar _ var) start end stmts)
+    | start <= end = mapM_ iteration [start .. end]
+    | otherwise    = modify' $ insertVar var (Some $ fromInteger actualEnd)
```
We already know in that clause that `end > start`, so there isn't much point in doing `max start end`.

jakzale replied:
It is a bit of defensive programming.  There is nothing preventing the user of writing the following programs:
```
for i = -1 to 0 do end;  assert i == 0;
```
```
for i = 0 to -1 do end;  assert i == 0;
```
```
for i = 0 to 0 do end;  assert i == 0;
```

This ensures that all of the above have defined semantics.

effectfully replied:
Sure, but we already have two different clauses:

1. `start <= end`
2. `otherwise`, i.e. `start > end`

and so it's not necessary in the second clause to handle the `start <= end` case as it's already handled by the first clause.

jakzale replied:
Yep, I can see it now.

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -144,22 +146,37 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
-    => Env (SomeUniConst f) -> Statement f -> Expr f a -> m (UniConst f a)
-evalStatementUni env (ELet (UniVar _ var) def) rest = do
+evalProgramUni :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
+    => Env (SomeUniConst f) -> Program f -> m (Env (SomeUniConst f))
```
It might be useful downstream for `evalProgramUni` to return an entire environment, but we're not using this currently and we can always add it later. I propose to make this function return `m ()` rather than `m (Env (SomeUniConst f))` just so we're not blocked on something that is not currently necessary.

jakzale replied:
Hmm, we are using it in testing at the moment, albeit it does cause some issues as we need to fix our semantics of renamings.

effectfully commented on field/TinyLang/Field/Typed/Core.hs:
```diff
@@ -228,32 +257,55 @@ isTracked uniq x env =
+stmtVarSigs' sigs (EFor uniVar _ _ stmts) = ScopedVarSigs free $ insertUnique uniq sig bound where
+    UniVar uni (Var uniq name) = uniVar
+    sig = VarSig name uni
+    ScopedVarSigs free bound = foldr (flip stmtVarSigs') sigs (C.unStatements stmts)
```
Here's the bug, it seems. `foldr` processes statements in reverse order, so a bound variable may become added to the set of free variables. If you look at this line below:

```haskell
exprVarSigs' sigs (EIf b x y) = exprVarSigs' (exprVarSigs' (exprVarSigs' sigs b) x) y
```

then it becomes clear that we have a left fold there (although there it's not critical unlike here), so `foldl'` should be used here as well.

jakzale replied:
Ah, I remember looking at that line, and the issue with sticking of `Expr f a` in a fold, is that `a` might be different, therefore it will not be possible to easily stuck them in a list, so I am not sure they can be stuck in a foldable.

effectfully replied:
Yeah, I'm not suggesting doing that, just giving an example.

jakzale replied:
Fixed now.

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -420,42 +506,23 @@ instance (KnownUni f a, Field f, Arbitrary f) => Arbitrary (Expr f a) where
+    shrink (ProgramWithEnv prog (Env vals)) =
+        -- TODO:  Check why is this keeping extra variable bindings in the env
+        flip map (shrink prog) $ \shrunk ->
+            ProgramWithEnv shrunk . Env . IntMap.intersection vals . unEnv $ progFreeVarSigs prog
```
Here's one mistake. You need to take `progFreeVarSigs` of the shrunk program, not the original one. Previously it was `exprFreeVarSigns expr` where `expr` comes from `shrunk`.

effectfully replied:
Yep, if fixed, then the shrunk test case looks correctly.

And if we define `shrink` as `const []`, then we get

```
        INITIAL STATE
          Env {unEnv = fromList [(1,Some 2),(7,Some 3),(9,Some (UniConst Bool False)),(23,Some 0)]}
        
        ORIGINAL PROGRAM
          assert (16 / t_7) >= (16 / t_7);
          let q_23 = if ?b_9 then y_1 else q_23;
``` 

which correctly generates values only for free variables.

jakzale replied:
Thanks!

jakzale replied:
Fixed.

jakzale commented on field/TinyLang/Field/Typed/Core.hs:
```diff
@@ -197,63 +227,100 @@ instance Eq f => Eq (Expr f a) where
+    . _scopedVarSigsFree
+    . execSVS
+    . traverse_ stmtVS
+
```
@effectfully, I've cleaned up the implementation of collecting `VarSig`s throughout the program.  What do you think?

effectfully replied:
I spent a bit of time thinking about this whole thing and I think there's a major problem with it (both the old version and the new one): it relies on global uniqueness of variables. Otherwise `x_0` in `let x_0 = x_0;` is both bound and free, and we do have programs like that. In the compiler I rename the term, so there it's fine. In old `Generator.hs` it should also be fine as we call `adjustUniquesForVars` there. But the requirement to have globally unique variables might be troubling in the new code.

So I think we should separate this procedure into two parts:

1. find free variables of a program (with variable declarations / binding external variables in `main` this won't require traversing the entire program) -- does not rely on global uniqueness, can be used in the new code freely
2. compute string names of all variables (regardless of whether they're free or bound) -- required by the compiler, relies on global uniqueness

jakzale replied:
Just to show off our new generators and shrinkers:  after enabling all the tests quick check found the case we are discussing and showed how it can break:
```
        ERROR
          evaluation of the renamed program did not match the evaluation of the original program

        INITIAL STATE
          Env {unEnv = fromList [(23,Some (UniConst Bool True))]}

        ORIGINAL PROGRAM
          let ?n_23 = ?n_23;

        ORIGINAL PROGRAM FINAL STATE
          Right (Env {unEnv = fromList [(23,Some (UniConst Bool True))]})

        RENAMED PROGRAM
          let ?n_24 = ?n_23;

        RENAMED PROGRAM FINAL STATE
          Right (Env {unEnv = fromList [(23,Some (UniConst Bool True)),(24,Some (UniConst Bool True))]})

        ProgramWithEnv (Program (Statements [ELet (UniVar {_uniVarUni = Bool, _uniVarVar = ?n_23}) (EVar (UniVar {_uniVarUni = Bool, _uniVarVar = ?n_23}))])) (Env {unEnv = fromList [(23,Some (UniConst Bool True))]})
        Use --quickcheck-replay=831249 to reproduce.
```

effectfully replied:
I wouldn't call it a breakage. Yes, final states are different, but that is fine as long as the final state of the renamed program corresponds to the final state of the original program, which is what we see here.

We could try extracting the final environment of the renamer and unrename the final state of the renamed program. Not sure, if it's going to work.

jakzale replied:
Fixed by returning a program trace.

effectfully commented on field/TinyLang/Field/Typed/Core.hs:
```diff
@@ -197,63 +227,100 @@ instance Eq f => Eq (Expr f a) where
+    ScopedVarSigs free bound <- get
+    let sig    = VarSig name uni
+        bound' = insertUnique uniq sig bound
+    put $ ScopedVarSigs free bound'
```
I'd use `modify` and a lens (or even `%=`) instead of `get` + `put`, but no big deal.

jakzale replied:
I am not that familiar with lenses.

Which lens library do you recommend:
- [lens](https://hackage.haskell.org/package/lens),
- [microlens](https://hackage.haskell.org/package/microlens), or
- [optics](https://hackage.haskell.org/package/optics)?

effectfully replied:
We use `lens` in the compiler. But it's not necessary to use a lens library, it's just that it saves some typing in cases like that.

jakzale replied:
Replaced with `modify`.  Need to look more at _lens_ to figure out how to use them.

jakzale commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -144,57 +187,75 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
+    => Program f -> EvalT f m ()
+evalProgramUni = evalStatementsUni . unProgram
+
+evalStatementsUni :: (Monad m, Eq f, Field f, AsInteger f)
```
@effectfully, I refactored the evaluator to a new transformer stack, per our discussion.

I was wondering if this method could be improved by using CPS.

Also, I will probably remove the `Statements f` wrapper, as it does not help anymore.

jakzale replied:
Done.

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -51,43 +56,85 @@ data EvalError f
-type MonadEvalError f m = MonadError (EvalError f) m
+-- | Evaluator Transformer Stack
+newtype EvaluatorT r s e (m :: Type -> Type) a =
+    EvaluatorT { runEvaluatorT :: ReaderT r (StateT s (ExceptT e m)) a }
```
Why introducing a generic version?

jakzale replied:
Yeah, I went a bit overboard wit refactoring.  Need to remove that.

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -51,43 +56,85 @@ data EvalError f
+    fromOption (throwError $ VariableNotInScopeEvalError var) $ lookupVar var env
 
-invEval :: (MonadEvalError g m, Field f) => f -> m f
+invEval :: (Monad m, Field f) => f -> EvalT g m f
```
Please change that back (and in other cases as well), we embed evaluation errors into compilation errors, so this has to be generic over `m`.

jakzale replied:
At this point `EvalT` implements `MonadEvalError`.  Just for clarification:  should I update all the exported functions to use the `mtl`-syle `MonadEvalError`?  (Or probably more aptly named `MonadEval`?)

effectfully replied:
What I mean is that functions that only throw and don't do anything with `Reader` or `State`, should run in the general `MonadEvalError`, because some of those functions are reused in the compiler. For example `invEval` doesn't do anything with `Reader` or `State`, hence it should stay generic and run in any `MonadEvalError`, which was defined as

```haskell
type MonadEvalError f m = MonadError (EvalError f) m
```

I.e. "any `m` is good enough as long as you can throw a `EvalError f` in it".

Functions that do stuff with `Reader` or `State` should stay the way they're implemented right now.

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -420,42 +512,23 @@ instance (KnownUni f a, Field f, Arbitrary f) => Arbitrary (Expr f a) where
+        vals <- genEnvFromVarSigs . progFreeVarSigs $ prog
+        return $ ProgramWithEnv prog vals
+    shrink (ProgramWithEnv prog (Env vals)) =
+        -- TODO:  Check why is this keeping extra variable bindings in the env
```
Outdated comment?

effectfully commented on test/Field/Textual.hs:
```diff
@@ -26,70 +31,73 @@ import           Test.Tasty.QuickCheck
 forgetStatementIDs (EAssert expr) = EAssert $ forgetIDs expr
+forgetStatementIDs (EFor uvar start end stmts) =
+    EFor (forgetID uvar) start end stmts' where
+        stmts' = forgetStatementIDs <$> stmts
```
I wouldn't bother introducing `stmts'` and would just inline its body, but not important.

jakzale replied:
By `pattern` trick you mean using the `ViewPatterns`?

effectfully commented on test/Field/Textual.hs:
```diff
@@ -98,47 +106,105 @@ test_printerParserRoundtrip =
-
-test_forLoops :: TestTree
-test_forLoops = parsePrintGolden "forLoops" $ unlines
-    [ "for i = 1 to 2 do"
```
Did we lose that test?

jakzale replied:
Refactored to its own file:
https://github.com/effectfully/tiny-lang/blob/8e36c56237619a82ed2d1c2a70b230b87ac029be/test/Field/golden/00-for-loops.field

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -144,57 +187,75 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
+            defR <- evalExprUni def
+            -- Record the store in the
+            -- TODO:  Fix a potential O(n^2) overhead
+            modify (++[Some defR])
```
I'd just collect the trace in reverse order, i.e.

```haskell
modify (Some defR :)
```

because we only care about traces staying the same after renaming, so their order doesn't matter.

Or you could use `DList`.

Commit:
[AST] fix statement vs expression

Fixes #51, #52

We make the following changes to the AST by adding
- a newly introduced a top-level `Program` node, which is parent to
- a newly introduced `Statements` node, which is parent to
- a `Statement` node, which is chaned to be a parent of
- an `Expr` node.

We make the following changes to Statement by
- introduce a `EFor` node to the typed AST.

We update update the evaluator by
- providing an explicit transformer stack (`EvalT`),
- fixing the semantics for for loops, and
- switching to CPS both in evaluator and normaliser.

We also add minor improvements to the codebase by
- fixing `Sign` vs `Sig` naming issue,
- cleaning up some naming conventions, and
- providing `stack bench` benchmark for testing generators for the new
  AST.

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -144,57 +190,72 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
+    exprR <- evalExpr expr
+    unless exprR $ throwError $ AssertionFailedEvalError expr
+    kont
+evalStatement (EFor uniVar@(UniVar _ var) start end body) kont
+    -- TODO:  Figure out if this is the right semantics
+    | start < end =
+        withVar var (Some $ fromIntegral start) $
+            evalStatements body $
+                evalStatement (EFor uniVar (start + 1) end body) kont
+    | otherwise =
+        let actualEnd = max start end in
+        withVar var (Some $ fromIntegral actualEnd) kont
```
Wait, so we have python-like loops now? I.e. `for i = x to y` now means `for i in [x, y)`. That's not what they have in the Sapling spec and it's currently the single source of truth for us. Or am I missing something?

jakzale replied:
Sorry, fixed this now.

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -204,72 +265,97 @@ denoteUniConst (UniConst Vector v) = unAField . fromInteger . packPositiveAsc $
+normStatements (Statements (s:rest)) kont = normStatement s $ \case
+          Nothing -> normStatements (Statements rest) kont
+          Just sN -> normStatements (Statements rest) $ \ (Statements restN) ->
+              kont $ Statements $ sN : restN
```
You could avoid all this wrapping-unwrapping by using the worker-wrapper transformation and having a local function like

```haskell
go :: [Statement f] -> ([Statement f] -> EvalT f m r) -> EvalT f m r
```

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -204,72 +265,97 @@ denoteUniConst (UniConst Vector v) = unAField . fromInteger . packPositiveAsc $
+      EConst (UniConst _ False) -> throwError $ AssertionFailedEvalError expr
+      EConst (UniConst _ True) -> kont Nothing
+      _ -> kont . Just $ EAssert exprN
+normStatement (EFor uniVar@(UniVar _ var) start end body) kont =
+    normStatements body $ \bodyN ->
+        let actualEnd = max start end in
+        withVar var (Some $ fromIntegral actualEnd) $ case bodyN of
+            Statements [] -> kont Nothing
+            _             -> kont . Just $ EFor uniVar start end bodyN
```
Technically speaking, normalizing a loop amounts to unrolling it. I don't know if we want to unroll the loop here, but please add a comment that we could.

Do you need this big-step normalization for something (apart from tests in `Textual.hs`)? I think I was going to use it for the shrinker, but now that we have a small-step normalizer there, we may not need the big-step one. I don't use normalization in the compiler.

Let's keep the code in this PR and remove it later, if we decide we don't need it.

Commit:
Refactor code to remove wrapping

Commit:
(WIP) Fix the semanics for for loop

Commit:
Fix indentation

Commit:
Fix whitespace

jakzale commented:
I suspect there will be a test error, but it will be tied to #76 due to lack of assignments and proper typing environments.

effectfully commented on field/TinyLang/Field/Evaluator.hs:
```diff
@@ -144,57 +190,70 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
+        where
+            go i fkont = withVar var (Some $ fromIntegral i) $
+                            evalStatements body fkont
+            actualEnd = Some $ fromIntegral $ max start end
```
As per previous discussion we already know in the `otherwise` clause that `start > end` and hence can do `withVar var start kont` there instead, but not a big deal.

effectfully reviewed (approved):
Ok, let's get this in!
